release-23.1: roachtest: add UploadCockroach and UploadWorkload APIs

renatolabs opened the pull request:
Backport:
  * 2/2 commits from "roachtest: add `UploadCockroach` and `UploadWorkload` APIs" (#115061)
  * 1/1 commits from "roachtest: indicate when version is too old in `UploadWorkload`" (#115447)

Please see individual PRs for details.

/cc @cockroachdb/release

****

Release justification: test only changes.

Commit:
roachtest: add `UploadCockroach` and `UploadWorkload` APIs

This commit updates the `clusterupgrade` package to provide functions
to allow tests to upload arbitrary versions of `cockroach` and
`workload` binaries.

Previously, the `UploadVersion` function was focused only on cockroach
binaries. With recent focus in upgrade testing, we would like to be
able to stage different versions of the `workload` binary during tests
as well. This will be especially useful for the `schemachange`
workload, since it only supports interactions between the current and
the previous version.

Fixes: #114518.

Release note: None

Commit:
roachtest: use `UploadWorkload` API in acceptance/version-upgrade

This commit makes use of the recently introduced `UploadWorkload` API
to upload a `workload` binary for a specific version. In this test, we
use a version that should be compatible with the upgrade currently
being performed.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/115433)
<!-- Reviewable:end -->

Commit:
roachtest: indicate when version is too old in `UploadWorkload`

In #115061, a `UploadWorkload` API was added to allow tests to upload
"arbitrary" versions of the `workload` binary, especially useful in
upgrade testing.

However, we only have `workload` binaries for v22.2+. In this commit,
we add a `bool` return value to `UplodWorkload` indicating whether a
binary was actually uploaded. Callers are expected to check this value
before attempting to use a workload binary.

The `acceptance/version-upgrade` test is updated in this commit in the
way described.

Epic: none

Release note: None

renatolabs commented:
Brought #115447 in.
